File: sirius-core/src/main/java/org/starcoin/core/BlockAddress.java
package org.starcoin.core;

import com.google.common.base.Preconditions;
import com.google.protobuf.ByteString;
import org.apache.commons.lang3.RandomUtils;
import org.starcoin.io.BlockOutputStream;
import org.starcoin.proto.Starcoin.ProtoBlockAddress;
import org.starcoin.util.HashUtil;
import org.starcoin.util.Utils;

import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.security.PublicKey;
import java.util.Arrays;

/**
 * BlockMsg Address:
 *
 * <p>RIPEMD160(sha256(public key))
 *
 * <p>pubkey is 65 bytes ecdsa public key result is 20 bytes hash
 *
 * @author Tim
 */
public final class BlockAddress implements Protobufable<ProtoBlockAddress>, ChainEncoder, Hashable {

    public static final int LENGTH = 20;
    public static final BlockAddress DEFAULT_ADDRESS =
            new BlockAddress(
                    new byte[]{
                            1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
                            11, 12, 13, 14, 15, 16, 17, 18, 19, 20
                    });

    // TODO add network flag to address
    private BlockAddress(byte[] addr) {
        Preconditions.checkArgument(addr.length == LENGTH, "expect address length:" + LENGTH);
        this.address = addr;
    }

    public BlockAddress() {
    }

    private byte[] address;

    public byte[] toBytes() {
        // not changeable
        return address.clone();
    }

    @Override
    public int getMarshalSize() {
        return LENGTH;
    }

    @Override
    public void writeTo(BlockOutputStream out) throws IOException {
        out.write(address);
    }

    public static BlockAddress readFrom(InputStream in) throws IOException {
        byte[] ad = new byte[LENGTH];
        int len = in.read(ad);
        Preconditions.checkArgument(
                len == LENGTH, new EOFException(BlockAddress.class.getName() + " expect more data"));
        return new BlockAddress(ad);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        BlockAddress that = (BlockAddress) o;
        return Arrays.equals(address, that.address);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(address);
    }

    @Override
    public String toString() {
        return Utils.HEX.encode(this.address);
    }

    public ProtoBlockAddress toProto() {
        return ProtoBlockAddress.newBuilder().setAddress(ByteString.copyFrom(address)).build();
    }

    @Override
    public Hash hash() {
        return Hash.of(this.address);
    }

    public static BlockAddress valueOf(ProtoBlockAddress address) {
        return new BlockAddress(address.getAddress().toByteArray());
    }

    public static BlockAddress valueOf(String addressHex) {
        Preconditions.checkNotNull(addressHex, "addressHex");
        return new BlockAddress(Utils.HEX.decode(addressHex));
    }

    public static BlockAddress random() {
        return new BlockAddress(RandomUtils.nextBytes(LENGTH));
    }

    public static BlockAddress genBlockAddressFromPublicKey(PublicKey publicKey) {
        return new BlockAddress(HashUtil.hash160(HashUtil.sha256(publicKey.getEncoded())));
    }

    public byte[] getAddress() {
        return address;
    }

    public void setAddress(byte[] address) {
        this.address = address;
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/BlockAddress.kt
package org.starcoin.sirius.core


import com.google.common.base.Preconditions
import com.google.protobuf.ByteString
import kotlinx.serialization.*
import org.apache.commons.lang3.RandomUtils
import org.starcoin.sirius.crypto.CryptoService
import org.starcoin.sirius.serialization.BinaryDecoder
import org.starcoin.sirius.serialization.BinaryEncoder
import org.starcoin.sirius.util.Utils
import java.security.PublicKey

@Serializable
class BlockAddress private constructor(private val bytes: ByteArray) : CachedHash() {

    init {
        Preconditions.checkArgument(bytes.size == LENGTH, "expect address length:$LENGTH")
    }

    fun toBytes() = bytes.copyOf()

    fun toByteString(): ByteString {
        return ByteString.copyFrom(this.bytes)
    }

    override fun toString(): String {
        return Utils.HEX.encode(this.bytes)
    }

    override fun hashData(): ByteArray {
        return this.bytes
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BlockAddress) return false

        if (!bytes.contentEquals(other.bytes)) return false

        return true
    }

    override fun hashCode(): Int {
        return bytes.contentHashCode()
    }

    @Serializer(forClass = BlockAddress::class)
    companion object : KSerializer<BlockAddress> {

        val LENGTH = 20

        val DUMMY_ADDRESS:BlockAddress by lazy { CryptoService.getDummyCryptoKey().getAddress() }

        @Deprecated("use DUMMY_ADDRESS")
        val DEFAULT_ADDRESS = DUMMY_ADDRESS

        override fun deserialize(input: Decoder): BlockAddress {
            return when (input) {
                is BinaryDecoder -> wrap(input.decodeByteArray())
                else -> wrap(input.decodeString())
            }
        }

        override fun serialize(output: Encoder, obj: BlockAddress) {
            when (output) {
                is BinaryEncoder -> output.encodeByteArray(obj.bytes)
                else -> output.encodeString(obj.toString())
            }
        }

        fun wrap(addressHex: String): BlockAddress {
            Preconditions.checkNotNull(addressHex, "addressHex")
            return BlockAddress(Utils.HEX.decode(addressHex))
        }

        fun wrap(address: ByteArray): BlockAddress {
            return BlockAddress(address)
        }

        fun wrap(address: ByteString): BlockAddress {
            Preconditions.checkNotNull(address, "address")
            Preconditions.checkArgument(!address.isEmpty, "address")
            return BlockAddress(address.toByteArray())
        }

        fun random(): BlockAddress {
            return BlockAddress(RandomUtils.nextBytes(LENGTH))
        }

        fun getAddress(publicKey: PublicKey): BlockAddress {
            return CryptoService.getAddress(publicKey)
        }
    }
}
